syntax = "proto3";

package blueOceanField.feature.process;

message FeatureProcess {
  string class_name = 1;
  map<string, FeatureProcessParameter> parameters = 2;
}

message FeatureProcessParameter {
  string name = 1;
  string description = 2;
  string type = 3;
  ParameterValue value = 4;
  repeated ConstraintInfo constraints = 5;
}

message ParameterValue {
  oneof kind {
    FeatureProcessParameter nested = 1;
    int32 int_value = 2;
    float float_value = 3;
    string string_value = 4;
    bool bool_value = 5;
    ListValue list_value = 6;
    MapValue map_value = 7;
  }
}

message ListValue {
  repeated ParameterValue items = 1;
}

message MapValue {
  map<string, ParameterValue> items = 1;
}

enum Constraint {
  CONSTRAINT_UNSPECIFIED = 0;
  REGEX = 1;
  UPPER_LIMIT = 2;
  LOWER_LIMIT = 3;
  INCLUSION = 4;
  EXCLUSION = 5;
  LENGTH = 6;
  UNIQUE = 7;
}

enum ConstraintTarget {
  TARGET_UNSPECIFIED = 0;
  SELF =1;
  KEY = 2;
  VALUE = 3;
}

message ConstraintInfo {
  Constraint type = 1;
  ConstraintTarget target = 2;
  repeated string args = 3;
}
